File: app/src/main/java/com/weathercock/cleanarchitecture/data/repository/CityRepositoryImpl.java
package com.weathercock.cleanarchitecture.data.repository;

import android.content.Context;

import com.weathercock.cleanarchitecture.data.dao.CityDao;
import com.weathercock.cleanarchitecture.data.entity.city.CityEntity;
import com.weathercock.cleanarchitecture.data.entity.city.PrefEntity;
import com.weathercock.cleanarchitecture.domain.model.city.CityModel;
import com.weathercock.cleanarchitecture.domain.repository.CityRepository;

import org.json.JSONException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * City Repository Implement
 * <p>
 * Created by tamura_k on 2016/05/27.
 */
public class CityRepositoryImpl implements CityRepository {

    private Context context;

    public CityRepositoryImpl(Context context) {
        this.context = context;
    }

    public List<CityModel> findAll() {
        List<CityModel> cityModelList = new ArrayList<>();

        try {
            CityDao cityDao = new CityDao(context);
            List<PrefEntity> prefEntityList = cityDao.find();
            for (PrefEntity prefEntity : prefEntityList) {
                String prefTitle = prefEntity.getTitle();
                for (CityEntity cityEntity : prefEntity.getCityList()) {
                    CityModel cityModel = new CityModel();
                    cityModel.setId(cityEntity.getId());
                    cityModel.setName(prefTitle + " " + cityEntity.getTitle());
                    cityModelList.add(cityModel);
                }
            }
        } catch (IOException | JSONException e) {
            e.printStackTrace();
        }

        return cityModelList;
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/data/repository/city/CityRepositoryImpl.kt
package com.kazakago.cleanarchitecture.data.repository.city

import android.content.Context
import com.kazakago.cleanarchitecture.data.file.dao.city.CityDao
import com.kazakago.cleanarchitecture.data.file.mapper.city.CityMapper
import com.kazakago.cleanarchitecture.domain.model.city.City
import com.kazakago.cleanarchitecture.domain.repository.city.CityRepository
import io.reactivex.Observable
import java.io.IOException

class CityRepositoryImpl(private val context: Context) : CityRepository {

    @Throws(IOException::class)
    override fun findAll(): Observable<City> {
        val cityDao = CityDao(context)
        val cityList = cityDao.find().flatMap { CityMapper.map(it) }
        return Observable.fromIterable(cityList)
    }

}
